### PR TITLE
ENH: Add pre-commit config file

### PR DESCRIPTION
**Pull request by fepegar**
Related to:
- #298
- #342
- #345
- #346
- #347
- #348
- #349

TODO:
- [x] Add `pre-commit` to requirements
- [ ] Add installation instructions in documentation, maybe in #298?
- [ ] Add `pre-commit` bot: https://pre-commit.ci. I will ask Microsoft to approve the app first

### CODE REVIEW

**Comment by fepegar** [comment]
Where do you think the requirement should be specified?

**Comment by ant0nsc** [comment]
> Where do you think the requirement should be specified?

sorry, I don't get your question?

**Comment by fepegar** [comment]
> sorry, I don't get your question?

The `pre-commit` library must be pip-installed for development, but I'm not sure where the requirement (e.g., `pre-commit==2.19.0`) should be specified. I mean, in which environment YAML file or requirements txt file within this repository I should add that line. Does that make sense?

**Comment by ant0nsc** [comment]
> The `pre-commit` library must be pip-installed for development, but I'm not sure where the requirement (e.g., `pre-commit==2.19.0`) should be specified. I mean, in which environment YAML file or requirements txt file within this repository I should add that line. Does that make sense?

Got it. Unfortunately in multiple places, root level `test_requirements.txt`, and `requirements_test.txt` and `environment.yml` in histo folder.

**Comment by vale-salvatelli** [comment]
Each package has its own flake8/mypy.ini, therefore each package should have their own pre-commit config file. 

Related to that, make sure the config points to our own defined flake8.ini to avoid inconsistencies.

**Comment by vale-salvatelli** [comment]
Each package has its own flake8/mypy.ini, therefore each package should have their own pre-commit config file. 

Related to that, make sure the config points to our own defined flake8.ini to avoid inconsistencies.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+- repo: https://github.com/PyCQA/flake8
+  rev: 4.0.1
+  hooks:
+  - id: flake8
+
+- repo: https://github.com/pre-commit/mirrors-autopep8
+  rev: v1.5.7
+  hooks:
+  - id: autopep8
```
Comment (vale-salvatelli) [inline .pre-commit-config.yaml:30]: why not adding also a hook for mypy?

Comment (vale-salvatelli) [inline .pre-commit-config.yaml:30]: why not adding also a hook for mypy?

Comment (fepegar) [reply .pre-commit-config.yaml:30]: We can. @ant0nsc feared that it might make things slow.

**Comment by fepegar** [comment]
Given that all the flake8 files are identical and that adding multiple configs in the repo doesn't feel right in terms on maintainability etc., I suggest we start with this simple setup and move on to different ones if necessary.

**Comment by vale-salvatelli** [comment]
> Given that all the flake8 files are identical and that adding multiple configs in the repo doesn't feel right in terms on maintainability etc., I suggest we start with this simple setup and move on to different ones if necessary.

seems reasonable if the files are identical. Let's just make sure it points towards at least one flake8.ini? I don't see any in the root where the config is currently placed

File: .pre-commit-config.yaml (hunk 1 above)
Comment (fepegar) [reply .pre-commit-config.yaml:30]: As below, I suggest leaving it out for now and adding it later on if we want to.

**Comment by fepegar** [comment]
> seems reasonable if the files are identical. Let's just make sure it points towards at least one flake8.ini? I don't see any in the root where the config is currently placed

Addressed in e34c759.

**Review by ant0nsc — APPROVED** [review]

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+# The structure of this was suggested by the author of pre-commit and maintainer of flake8
+# See https://stackoverflow.com/a/66485642/3956024
+- repo: https://github.com/PyCQA/flake8
+  rev: 4.0.1
+  hooks:
+  - id: flake8
+    name: flake8 ./hi-ml/
+    alias: flake8-hi-ml
+    files: ^hi-ml/
```
[the remark this answers is not in the document]
Comment (fepegar) [reply .pre-commit-config.yaml:30]: Doesn't that happen by default?

**Review by vale-salvatelli — APPROVED** [review]

File: .pre-commit-config.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (fepegar) [reply .pre-commit-config.yaml:30]: Resolving this as I'm pretty sure flake8 won't look at other types of files.

**Pull request merged**